app/src/main/java/info/nightscout/androidaps/utils/stats/TddCalculator.kt

```diff
@@ -161,6 +161,102 @@ class TddCalculator @Inject constructor(
         aapsLogger.debug(LTag.CORE, tdd.toString())
         return tdd
     }
+    fun calculate4Daily():TotalDailyDose {
+        val  startTime = dateUtil.now() - T.hours(hour = 4).msecs()
+        val endTime = dateUtil.now()
+        val tdd = TotalDailyDose(timestamp = startTime)
+        //val result = TotalDailyDose()
+        repository.getBolusesDataFromTimeToTime(startTime, endTime, true).blockingGet()
+            .filter { it.type != Bolus.Type.PRIMING }
+            .forEach { t ->
+                //val midnight = MidnightTime.calc(t.timestamp)
+                //val tdd = result[midnight] ?: TotalDailyDose(timestamp = midnight)
+                tdd.bolusAmount += t.amount
+                //result.put(midnight, tdd)
+            }
+        repository.getCarbsDataFromTimeToTimeExpanded(startTime, endTime, true).blockingGet().forEach { t ->
+            //val midnight = MidnightTime.calc(t.timestamp)
+            //val tdd = result[midnight] ?: TotalDailyDose(timestamp = midnight)
+            tdd.carbs += t.amount
+            //result.put(midnight, tdd)
+        }
+        val calculationStep = T.mins(5).msecs()
+        val tempBasals = iobCobCalculator.getTempBasalIncludingConvertedExtendedForRange(startTime, endTime, calculationStep)
+        for (t in startTime until endTime step calculationStep) {
+
+            //val midnight = MidnightTime.calc(t)
+            //val tdd = result[midnight] ?: TotalDailyDose(timestamp = midnight)
+            val tbr = iobCobCalculator.getTempBasalIncludingConvertedExtended(t)
+            val profile = profileFunction.getProfile(t) ?: continue
+            val absoluteRate = tbr?.convertedToAbsolute(t, profile) ?: profile.getBasal(t)
+            tdd.basalAmount += absoluteRate / 60.0 * 5.0
+
+            if (!activePlugin.activePump.isFakingTempsByExtendedBoluses) {
+                // they are not included in TBRs
+                val eb = iobCobCalculator.getExtendedBolus(t)
+                val absoluteEbRate = eb?.rate ?: 0.0
+                tdd.bolusAmount += absoluteEbRate / 60.0 * 5.0
+            }
+            //result.put(midnight, tdd)
+        }
+        //for (i in 0 until tdd.size()) {
+        //val tdd = result.valueAt(i)
+        tdd.totalAmount = tdd.bolusAmount + tdd.basalAmount
+        //}
+
+
+        aapsLogger.debug(LTag.CORE, tdd.toString())
+        return tdd
+    }
+
+
+    fun calculate8Gap():TotalDailyDose {
```

Comment:
this is copy-paste
replace with 1 function with params

Reply:
If you allow this to be used, then I can try and work that out, but right now that will take me some time, and I'd like to get the updated model into wider circulation.